### PR TITLE
Fix activator utils when using with classes containig struct default parameter value

### PR DESCRIPTION
**Pull request by pakrym**
Fixes: https://github.com/aspnet/Home/issues/2072
Test: https://github.com/aspnet/DependencyInjection/pull/538

### CODE REVIEW

**Comment by pakrym** [comment]
Also filed https://github.com/aspnet/Common/issues/245

**Comment by pakrym** [comment]
Same thing here, github does not pickup new commits: https://github.com/aspnet/Common/commits/pakrym/struct-default-test

**Comment by pakrym** [comment]
Fix got pretty ugly because `HasDefaultValue` throws for `DateTime`

**Comment by pakrym** [comment]
@pranavkm  Now I'm thinking about extracting utility method for getting default value so it can be reused not only in DI.

**Comment by pranavkm** [comment]
> Now I'm thinking about extracting utility method for getting default value so it can be reused not only in DI.

Seems alright  - but it's like 10 lines of code. Not sure, it would warrant a separate package of it's own.

**Comment by pakrym** [comment]
Source package?

**Comment by pranavkm** [comment]
Ask around - maybe @rynowak \ @davidfowl have strong opinions about it. I'm ok with a source package, but it does some like overkill for a small amount of utility code.

File: shared/Microsoft.Extensions.ParameterHelper.Sources/ParameterHelper.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+
+namespace Microsoft.Extensions.Internal
+{
+    internal class ParameterHelper
```
Comment (pranavkm) [inline shared/Microsoft.Extensions.ParameterHelper.Sources/ParameterHelper.cs]: `ParameterDefaultValue`

File: shared/Microsoft.Extensions.ParameterHelper.Sources/ParameterHelper.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
```
Comment (pranavkm) [inline shared/Microsoft.Extensions.ParameterHelper.Sources/ParameterHelper.cs]: Probably don't need this.

File: shared/Microsoft.Extensions.ActivatorUtilities.Sources/sharedsources.props [hunk 3]
```diff
@@ -0,0 +1,8 @@
+<Project>
```
Comment (pranavkm) [inline shared/Microsoft.Extensions.ActivatorUtilities.Sources/sharedsources.props]: cc @natemcmaster

File: shared/Microsoft.Extensions.ParameterHelper.Sources/ParameterHelper.cs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+            var tryToGetDefaultValue = true;
+            defaultValue = null;
+
+            try
+            {
+                hasDefaultValue = parameter.HasDefaultValue;
+            }
+            catch (FormatException) when (parameter.ParameterType == typeof(DateTime))
+            {
+                // TODO: workaround https://github.com/dotnet/corefx/issues/12338
```
Comment (pranavkm) [inline shared/Microsoft.Extensions.ParameterHelper.Sources/ParameterHelper.cs]: I don't think there's any TODO here?

File: shared/Microsoft.Extensions.ParameterHelper.Sources/ParameterHelper.cs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+            }
+
+            if (hasDefaultValue)
+            {
+                if (tryToGetDefaultValue)
+                {
+                    defaultValue = parameter.DefaultValue;
+                }
+
+                // TODO: workaround for https://github.com/dotnet/corefx/issues/11797
```
Comment (pranavkm) [inline shared/Microsoft.Extensions.ParameterHelper.Sources/ParameterHelper.cs]: Nor here.

File: NuGetPackageVerifier.json [hunk 6]
```diff
@@ -8,6 +8,7 @@
       "Microsoft.Extensions.CopyOnWriteDictionary.Sources": {},
       "Microsoft.Extensions.HashCodeCombiner.Sources": {},
       "Microsoft.Extensions.ObjectMethodExecutor.Sources": {},
+      "Microsoft.Extensions.ParameterHelper.Sources": {},
```
Comment (pranavkm) [inline NuGetPackageVerifier.json]: `ParameterDefaultValue.Sources` or something like that.

**Pull request merged**